flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/JobScopedResourceTracker.java

```diff
@@ -0,0 +1,225 @@
+
+	JobScopedResourceTracker(JobID jobId) {
+		this.jobId = Preconditions.checkNotNull(jobId);
+	}
+
+	public void notifyResourceRequirements(Collection<ResourceRequirement> newResourceRequirements) {
+		Preconditions.checkNotNull(newResourceRequirements);
+
+		resourceRequirements.clear();
+		for (ResourceRequirement newResourceRequirement : newResourceRequirements) {
+			resourceRequirements.incrementCount(newResourceRequirement.getResourceProfile(), newResourceRequirement.getNumberOfRequiredSlots());
+		}
+		findExcessSlots();
+		tryAssigningExcessSlots();
+	}
+
+	public void notifyAcquiredResource(ResourceProfile resourceProfile) {
+		Preconditions.checkNotNull(resourceProfile);
+		final Optional<ResourceProfile> matchingRequirement = findMatchingRequirement(resourceProfile);
+		if (matchingRequirement.isPresent()) {
+			resourceToRequirementMapping.incrementCount(matchingRequirement.get(), resourceProfile, 1);
+		} else {
+			LOG.debug("Job {} acquired excess resource {}.", resourceProfile, jobId);
+			excessResources.incrementCount(resourceProfile, 1);
+		}
+	}
+
+	private Optional<ResourceProfile> findMatchingRequirement(ResourceProfile resourceProfile) {
+		for (Map.Entry<ResourceProfile, Integer> requirementCandidate : resourceRequirements.getResourceProfilesWithCount().entrySet()) {
+			ResourceProfile requirementProfile = requirementCandidate.getKey();
+
+			// beware the order when matching resources to requirements, because ResourceProfile.UNKNOWN (which only
+			// occurs as a requirement) does not match any resource!
+			if (resourceProfile.isMatching(requirementProfile) && requirementCandidate.getValue() > resourceToRequirementMapping.getNumFulfillingResources(requirementProfile)) {
+				return Optional.of(requirementProfile);
+			}
+		}
+		return Optional.empty();
+	}
+
+	public void notifyLostResource(ResourceProfile resourceProfile) {
+		Preconditions.checkNotNull(resourceProfile);
+		if (excessResources.getResourceCount(resourceProfile) > 0) {
+			LOG.trace("Job {} lost excess resource {}.", jobId, resourceProfile);
+			excessResources.decrementCount(resourceProfile, 1);
+			return;
+		}
+
+		Set<ResourceProfile> fulfilledRequirements = resourceToRequirementMapping.getRequirementsFulfilledBy(resourceProfile).keySet();
+
+		if (!fulfilledRequirements.isEmpty()) {
+			// determine for which of the requirements, that the resource could be used for, the resource count should be reduced for
+			ResourceProfile assignedRequirement = null;
+
+			for (ResourceProfile requirementProfile : fulfilledRequirements) {
+				assignedRequirement = requirementProfile;
+
+				// try finding a requirement that has too many resources; if non are exceeding the requirements we deduct
+				// the resource from any requirement having such a resource
+				if (resourceRequirements.getResourceCount(requirementProfile) < resourceToRequirementMapping.getNumFulfillingResources(requirementProfile)) {
+					break;
+				}
```

Comment:
Should this ever happen? I thought that all slots which over fulfill the requirements will be added to the `excessResources` field.

Reply:
hmm...that is true.

Reply:
It was necessary at some point because we did not always searched for excess slots when requirements changed, but only when acquiring resources.